google/blockly
# fix: Make message files work in unpackaged mode, and rebuild msg files

## The basics




## The details
### Resolves

Fixes #6123 

### Proposed Changes

Updates the Msg files to use `Blockly.Msg` instead of `messages`, and associated changes in the preamble and wrapper to make that approach work in all cases.

n.b. Thanks to @cpcallen for suggesting the approach taken here.

#### Behavior Before Change

- Message loading works when importing as a module
- Message loading works when loading the wrapped message files through a script tag
- Message loading **does not work** when loading the unwrapped message files through a script tag

#### Behavior After Change

- Message loading works when importing as a module
- Message loading works when loading the wrapped message files through a script tag
- Message loading works when loading the unwrapped message files through a script tag

### Reason for Changes

In #6184 the message files were changed. Instead of assigning each translation directly on `Blockly.Msg`, the translations were put on a new object called `messages`. In the UMD wrapper for the language files, all of the properties of that new object are then placed on `Blockly.Msg`. This works for both of the first two cases above, but it doesn't work for anyone who is using the unwrapped message files.

That is a scenario we need to support for now since it is explicitly in our documentation that you can download and run the code from github.

Also, the file `blockly.min.js` in the packaged output does not use the wrapped message file, so e.g. using a script tag to load `<script src="https://unpkg.com/blockly"></script>` does not work because the unwrapped message files don't load prior to this change. That line of code is in the Getting Started codelab so that also needs to work. That will be fixed in a follow up PR.

### Test Coverage

- Playground in core loading `msg/messages.js` (default behavior. this is the raw message file that isn't run through the python script and does not have the umd wrapper so it is not a good test of any real world behavior and should probably be changed)
- Playground in core loading `'dist/msg/es.js'` instead (compiled through python script and umd wrapped)
- Playground in core loading `'msg/js/es.js'` instead (compiled through python script but not umd wrapped)
- scroll-options plugin linking to this build of blockly (verified through inspecting node_modules that it points to this build and not a random other plugin's old version of blockly...)
- scroll options plugin with these changes in `test/index.js`:

```js
import * as Fr from 'blockly/msg/fr';
import * as Es from 'blockly/msg/es';
// ...
function createWorkspace(blocklyDiv, options) {
  Blockly.setLocale(Fr);
// ...
}
```
- and with these changes:

```js
import * as Fr from 'blockly/msg/fr';
import * as Es from 'blockly/msg/es';
// ...
function createWorkspace(blocklyDiv, options) {
  Blockly.setLocale(Fr);
  Blockly.setLocale(Es);
// ...
}
```

- Getting Started codelab, replacing `<script src="https://unpkg.com/blockly"></script>` with `<script src="../../../../blockly/dist/blockly.min.js"></script>` (linked to this build of blockly. this is the equivalent file that gets loaded when visiting the unpkg link. it contains the unwrapped english message file as noted above.)
- Same as above but also adding `<script src="../../../../blockly/dist/msg/fr.js"></script>` (compiled and umd wrapped)
- Block factory demo in core using `msg/js/en.js` (default - compiled but not umd wrapped)
- Block factory demo in core using `dist/msg/az.js` (compiled and umd wrapped)
- Plane demo in samples, linked to this build of blockly
